feat(tree-view): added compact, bordered, and background transparent variants

mattnolting opened the pull request:
closes #4177 

[Previews start here](https://patternfly-pr-4242.surge.sh/components/tree-view#guides)

patternfly-build commented:
Preview: https://patternfly-pr-4242.surge.sh

mcoker commented on src/patternfly/components/TreeView/examples/TreeView.md:
```diff
@@ -716,9 +818,15 @@ beta: true
+| `.pf-m-compact` | `.pf-c-tree-view` | Modifies the tree view to the compact presentation. |
+| `.pf-m-background-transparent` | `.pf-c-tree-view.pf-m-compact` | Modifies the tree view compact variant node containers to have a transparent background. |
 | `.pf-m-current` | `.pf-c-tree-view__node` | Modifies the tree view node to be current. |
+| `.pf-m-truncate-text` | `.pf-c-tree-view` | Modifies the tree view text to truncate. |
```
Do this and `.pf-m-truncate` do the same thing, except this applies it to all text elements in the tree vieww, and `.pf-m-truncate` applies to an individual element? If so I wonder if we could use the same modifier name for both, similar to how we do on table?

mcoker reviewed (changes requested):
This looks amazing! Great job on the lines, too! Left some comments for review.

Also curious why these top-level nodes aren't buttons, but the children are?

### top level
<img width="625" alt="Screen Shot 2021-07-30 at 2 22 23 PM" src="https://user-images.githubusercontent.com/35148959/127706104-7c86e366-2320-4a89-92ee-bba83816995e.png">

### children
<img width="931" alt="Screen Shot 2021-07-30 at 2 22 02 PM" src="https://user-images.githubusercontent.com/35148959/127706122-d1da6ed4-bb65-4c94-9cbf-a70f487d0579.png">

mattnolting commented:
@mcoker updated

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -110,6 +310,18 @@ $pf-c-tree-view--MaxNesting: 10;
+.pf-c-tree-view__node-content {
+  display: flex;
+  flex-direction: column;
+  row-gap: var(--pf-c-tree-view__node-content--RowGap);
```
WDYT about this?

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +46,250 @@ $pf-c-tree-view--MaxNesting: 10;
-    > .pf-c-tree-view__content > .pf-c-tree-view__node > .pf-c-tree-view__node-toggle > .pf-c-tree-view__node-toggle-icon {
+    > .pf-c-tree-view__content > .pf-c-tree-view__node > .pf-c-tree-view__node-toggle > .pf-c-tree-view__node-toggle-icon,
+    > .pf-c-tree-view__content > .pf-c-tree-view__node > .pf-c-tree-view__node-container > .pf-c-tree-view__node-toggle > .pf-c-tree-view__node-toggle-icon {
       transform: rotate(var(--pf-c-tree-view__list-item--m-expanded__node-toggle-icon--Rotate));
```
WDYT about this?

mcoker reviewed (changes requested):
Looking really good. Left a few comments, and followed up on some from my last review - curious your feedback on those. 

Also I wonder if we should call the "bordered" variation something else. We use `.pf-m-bordered` in accordion and lists to create a border between items (similar to the border between items in the compact variant here). These are more like hierarchy lines than a border between things, though I'm not sure what to call it. @mcarrano wdyt?

Also the lines here don't align with the toggles like they do in the regular bordered variant. It looks like this matches the compact design, but not the compact/transparent-bg design (the lines appear to align with the toggles). I'm wondering if what we have here is OK. @mcarrano wdyt?

<img width="372" alt="Screen Shot 2021-08-05 at 9 53 33 AM" src="https://user-images.githubusercontent.com/35148959/128373275-ca51e2df-9542-496d-81f9-9485f629cb0f.png">
<img width="257" alt="Screen Shot 2021-08-05 at 9 53 27 AM" src="https://user-images.githubusercontent.com/35148959/128373278-5f620f4f-4fd3-4a7a-b831-3e93c425bf9b.png">
<img width="298" alt="Screen Shot 2021-08-05 at 9 53 22 AM" src="https://user-images.githubusercontent.com/35148959/128373280-4e73cc70-e6bf-418f-a6b0-d06e9c09c06a.png">
<img width="194" alt="Screen Shot 2021-08-05 at 9 53 16 AM" src="https://user-images.githubusercontent.com/35148959/128373282-c7aa450d-7e9d-443b-b689-7c304bec3be5.png">

mattnolting commented:
@mcarrano @mcoker `pf-m-bracketed`??

mcoker commented:
> Also, I don't see the alignment issue you are referencing above.

@mcarrano 

Here's the compact version. Note the vertical lines under an expandable node are offset to the left of the toggle.

<img width="315" alt="Screen Shot 2021-08-05 at 12 47 55 PM" src="https://user-images.githubusercontent.com/35148959/128397388-1e5f4348-5de8-4843-be1c-db6cc3f9c8a8.png">

This matches the design

<img width="393" alt="Screen Shot 2021-08-05 at 12 48 28 PM" src="https://user-images.githubusercontent.com/35148959/128397416-b0c0d8c4-0b82-4d0c-91e7-391f5aa85f04.png">

Here is the compact/no-grey-box version - note the lines are in the same position as the regular compact variation

<img width="435" alt="Screen Shot 2021-08-05 at 12 48 04 PM" src="https://user-images.githubusercontent.com/35148959/128397444-b525deaa-2354-48b7-8247-58633425275f.png">

But in the design, they aren't offset to the left - they seem to be centered with the toggle on the parent expandable node

<img width="472" alt="Screen Shot 2021-08-05 at 12 48 55 PM" src="https://user-images.githubusercontent.com/35148959/128397620-9ff6c701-d713-4cad-a6b1-a014731a8d3c.png">

Just curious about how it should be - is what's in the PR OK?

mcarrano commented:
@mattnolting I like "bracketed"

@mcoker now I see what you are saying about alignment. I guess this is OK since I didn't even notice the difference :) @doruskova what do you think?

mcoker commented:
@mattnolting @mcarrano bracketed works for me. Another thought - `pf-m-guided`, as representing line guides.

mattnolting commented:
@mcoker I can create a followup issue to track indentation.

doruskova commented:
@mcoker @mattnolting @mcarrano Wow! This looks really great! I think that the alignment looks good. It's not a big difference and I didn't notice it.

mattnolting commented:
@mcarrano I updated the modifier to `.pf-m-guides`, as in it contains guides. I think that would be a good precedent to set for anything that contains guides, rather than exclusively borders.

@mcoker @doruskova I updated the compact, no background variant padding and border positioning. I think it looks right/accurate now.

doruskova commented:
@mattnolting Looks good to me!

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-guides--border--ZIndex: var(--pf-c-tree-view--m-guides--base-border--ZIndex);
+
+  // Base
+  --pf-c-tree-view--m-guides--base-border--BorderColor: var(--pf-global--BorderColor--100);
```
just curious why these aren't named `--pf-c-tree-view--m-guides--border-color--base`, `--pf-c-tree-view--m-guides--border-width--base`, etc.

mcoker replied:
also if we keep `border` in the var name, should that be guide?

mattnolting replied:
udpated

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  // Compact
+  // ================================================================== //
+  // Base
+  --pf-c-tree-view--m-compact--base-border--Left--offset: var(--pf-global--spacer--md);
```
Curious why some are `--base-border--Left--offset` and others are `--indent--base`.

mattnolting replied:
Indent refers to the node padding while offset refers to the guide positioning. They separate because they change relatively as nesting increases past 2 levels deep and need to be calculated differently.

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-compact__node--indent--base: var(--pf-c-tree-view__node--indent--base); // based off of the expected width of the toggle button
+  --pf-c-tree-view--m-compact__node--nested-indent--base: var(--pf-global--spacer--lg);
+  --pf-c-tree-view--m-compact--border--Left: var(--pf-c-tree-view--m-compact--base-border--Left);
+  --pf-c-tree-view--m-compact--border--nested--Left--offset: calc(var(--pf-c-tree-view--m-compact__node--nested-indent--base) / 1.5);
```
doesn't appear to be in use

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-compact--base-border--Left: calc(var(--pf-c-tree-view__node--PaddingLeft) - var(--pf-c-tree-view--m-compact--base-border--Left--offset));
+  --pf-c-tree-view--m-compact__node--indent--base: var(--pf-c-tree-view__node--indent--base); // based off of the expected width of the toggle button
+  --pf-c-tree-view--m-compact__node--nested-indent--base: var(--pf-global--spacer--lg);
+  --pf-c-tree-view--m-compact--border--Left: var(--pf-c-tree-view--m-compact--base-border--Left);
```
should this be guide?

```suggestion
  --pf-c-tree-view--m-compact--guide--Left: var(--pf-c-tree-view--m-compact--base-border--Left);
```

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+
+  // Node
+  --pf-c-tree-view--m-compact__node--PaddingTop: 0;
+  --pf-c-tree-view--m-compact__node--PaddingRight: 0;
```
not in use

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-compact__list-item--last-child--before--Height: var(--pf-c-tree-view--m-compact__node--before--Top);
+  --pf-c-tree-view--m-compact__list-item--nested--before--Top: calc(var(--pf-c-tree-view__node--PaddingTop--base) * -1);
+  --pf-c-tree-view--m-compact__list-item--nested--last-child--before--Height: calc(var(--pf-c-tree-view--m-compact__node--before--Top) + var(--pf-c-tree-view__node--PaddingTop--base));
+  --pf-c-tree-view--m-compact__list-item--nested--last-child--before--Top: calc(var(--pf-c-tree-view__node--Padding) * -2);
```
not in use

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-compact--m-no-background--base-border--Left: calc(var(--pf-c-tree-view__node--PaddingLeft) - var(--pf-c-tree-view--m-compact--m-no-background--base-border--Left--offset));
+  --pf-c-tree-view--m-compact--m-no-background__node--indent--base: var(--pf-c-tree-view__node--indent--base); // based off of the expected width of the toggle button
+  --pf-c-tree-view--m-compact--m-no-background__node--nested-indent--base: var(--pf-global--spacer--2xl);
+  --pf-c-tree-view--m-compact--m-no-background--border--Left: var(--pf-c-tree-view--m-compact--m-no-background--base-border--Left);
```
not in use

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-compact--m-no-background__node--indent--base: var(--pf-c-tree-view__node--indent--base); // based off of the expected width of the toggle button
+  --pf-c-tree-view--m-compact--m-no-background__node--nested-indent--base: var(--pf-global--spacer--2xl);
+  --pf-c-tree-view--m-compact--m-no-background--border--Left: var(--pf-c-tree-view--m-compact--m-no-background--base-border--Left);
+  --pf-c-tree-view--m-compact--m-no-background--border--nested--Left--offset: calc(var(--pf-c-tree-view--m-compact--m-no-background__node--nested-indent--base) / 1.5);
```
not in use

mcoker commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+  --pf-c-tree-view--m-compact--m-no-background__node--nested--PaddingTop: 0;
+  --pf-c-tree-view--m-compact--m-no-background__node--nested--PaddingBottom: 0;
+  --pf-c-tree-view--m-compact--m-no-background__node--before--Top: calc(var(--pf-c-tree-view--m-compact__node-container--PaddingTop) + var(--pf-c-tree-view--m-compact--m-no-background__node--nested--PaddingTop) - #{pf-size-prem(4px)});
+  --pf-c-tree-view--m-compact--m-no-background__node--before--Top: calc(var(--pf-c-tree-view--m-compact__node-container--nested--PaddingTop) + var(--pf-c-tree-view--m-compact__node--nested--PaddingTop) + #{pf-size-prem(4px)});
```
duplicate vars. I'm assuming you want to keep this bottom one?

mcoker reviewed (changes requested):
Design looks great! Left some comments.

mcarrano reviewed (approved):
These look great @mattnolting !

mattnolting commented on src/patternfly/components/TreeView/tree-view.scss:
```diff
@@ -36,60 +47,322 @@ $pf-c-tree-view--MaxNesting: 10;
+        .pf-c-tree-view__node::before {
+          position: absolute;
+          left: var(--pf-c-tree-view--m-guides--border--Left);
+          z-index: var(--pf-c-tree-view--m-guides--border--ZIndex);
```
Just following the design, can we follow up in a separate issue?

mcoker reviewed (approved):
LGTM! Fantastic work with everything! Opened a follow up issue with a couple of small things here - https://github.com/patternfly/patternfly/issues/4277

patternfly-build commented:
:tada: This PR is included in version 4.127.0 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v4.127.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/4.127.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
